### PR TITLE
Stable layout

### PR DESCRIPTION
**Pull request by Lokathor**
TODO:
* [x] UnsafeCell / Cell when T: StableLayout
* [x] Fix reference types
* [x] Str slices (Closes https://github.com/Lokathor/chromium/issues/3)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -31,3 +31,56 @@ pub use c_shared_slice::*;
...
+unsafe impl StableLayout for i64 { }
+unsafe impl StableLayout for isize { }
+
+unsafe impl StableLayout for bool { }
+unsafe impl StableLayout for char { }
+unsafe impl StableLayout for () { }
+
+unsafe impl<T> StableLayout for core::num::Wrapping<T> where T: StableLayout { }
+
+unsafe impl<T> StableLayout for &T where T: Sized { }
```
Comment (thomcc) [inline src/lib.rs]: This, and the other references, are unsound, as if this is for rust<->rust, it's busted as soon as the other end dereferences it.

It should be `where T: Sized + StableLayout`.

Comment (Lokathor) [reply src/lib.rs]: Yeah I even had it require StableLayout in a previous version, dunno why i forgot to put both as the requirement

**Review by danielhenrymantilla — APPROVED** [review]
Looks good to me. A few nits:

  - `<T>` already implies `T : Sized`, but I don't mind it being extra explicit :+1:

  - you could avoid some code duplication by making `_str` be a `transparent/C` wrapper over `_slice<u8>`, given that the difference between the two does not involve layout.

  - I think that `PhantomData` is a null-layout thingy with the semantics of the wrappee. But since `StableLayout` is not about semantics but about layout, I'd remove the bound for it (null layout is always stable). That being said, you could then offer your own `PhantomUnstableLayout` if somebody were to need it (I struggle to see why, given that it is not an auto-trait, but we never know).

      - similarly, the impl for `[T; 0]` could remove the bound on `T` too.

  - Since very few people will need `&mut str`, it could be feature-gated.

  - you still interested in someone submitting a PR for the derive?

**Comment by Lokathor** [comment]
* I don't think that using a transparent wrapper would reduce the _str code by much.
* PhantomData is definitely null layout. However, in the same way that we're counting pointers/references as "stable layout" only if they point to stable layout data _just for extra safety_, I think we should stick to the extra limit on PhantomData as well. It would always be possible to reduce the limit later (possibly even via cargo feature if necessary).
* I know that `&mut str` is weird and rare, but I'd rather just be plain and have the coverage instead of making people fiddle with features. It's only like a 100 line file, not any big drag on compile times.
* I think since the derive lives in a separate crate (right?) someone can just make a separate repo, and then I'll point to them in the crate docs and readme once they're ready.

**Pull request merged**